### PR TITLE
[1LP][RFR] Ec2 sts assume role test

### PR DESCRIPTION
**Pull request by mmojzis**
QE_TEST_COVERAGE FOR BZ https://bugzilla.redhat.com/show_bug.cgi?id=1631845

{{ pytest:  cfme/tests/cloud/test_providers.py -k "test_add_ec2_provider_with_sts_assume_role" -v }}

### CODE REVIEW

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by john-dupuy** [comment]
@mmojzis marking this WIPTEST until the yamls MR is merged. Also, as far as I can see, this is only a semi-automated test, i.e. it has some manual setup required. So I expect in order for it to be merged it must be (a) fully automated or (b) a manual test that can be aided through some utility automation that lives in e.g. `cfme/utils`.

**Comment by mmojzis** [comment]
> @mmojzis marking this WIPTEST until the yamls MR is merged. Also, as far as I can see according to your `Requires` comment (and please correct me if I'm mistaken), this is a semi-automated test, i.e. it has some manual setup required. So I expect in order for it to be merged it must be (a) fully automated or (b) a manual test that can be aided through some utility automation that lives in e.g. `cfme/utils`.

Well, there is no way how to automate it.
I only need to have these resources created.
I did it manually and it doesn't have to be recreated again.
Test should work without any manual intervention.

**Comment by mmojzis** [comment]
Failed PRT runs are not related to this PR.

**Comment by john-dupuy** [comment]
@mmojzis Got it, I think I misunderstood that `Requires` statement -> that is on the EC2 side not the CFME side. Perhaps add a comment about that? 

Thanks for the clarification!

**Comment by mmojzis** [comment]
@john-dupuy https://github.com/ManageIQ/integration_tests/pull/9436/files#diff-4f10bc4cf82a0f08488c8079d4cd0085R1309 added comment here. I hope it's not confusing now.

File: cfme/tests/cloud/test_providers.py [hunk 1]
```diff
@@ -1297,3 +1300,65 @@ def test_add_ec2_provider_with_non_default_url_endpoint():
...
+                       secret=credentials[assume_role_creds]['password'])
+    endpoint = EC2Endpoint(
+        assume_role_arn=cfme_data['management_systems']['ec2west']['sts_assume_role']['role_arn'],
+        credentials=creds)
+
+    collection = appliance.collections.cloud_providers
+    prov = collection.instantiate(
+        prov_class=EC2Provider, name=fauxfactory.gen_alphanumeric(5),
+        region_name=cfme_data['management_systems']['ec2west']['region_name'],
+        endpoints=endpoint)
```
Comment (john-dupuy) [inline cfme/tests/cloud/test_providers.py]: I would expect this instantiation to be part of a fixture, with only L1364 being part of the test.

File: cfme/tests/cloud/test_providers.py [hunk 2]
```diff
@@ -1297,3 +1300,65 @@ def test_add_ec2_provider_with_non_default_url_endpoint():
...
+
+    request.addfinalizer(prov.delete_if_exists)
+    prov.create()
```
Comment (john-dupuy) [inline cfme/tests/cloud/test_providers.py]: Is there any verification that needs to be done after the provider is successfully added? Will the test error out or fail if the  provider is not successfully added?

File: cfme/tests/cloud/test_providers.py (hunk 1 above)
Comment (mmojzis) [reply cfme/tests/cloud/test_providers.py]: Reworked as a fixture.

File: cfme/tests/cloud/test_providers.py (hunk 2 above)
Comment (mmojzis) [reply cfme/tests/cloud/test_providers.py]: Well I wanted to add some smoke tests after adding this provider, but I think for now checking whether provider refresh was successful is sufficient.

**Comment by dajoRH** [comment]
I detected some fixture changes in commit bf18e5216005605b1f68649b52be4f3816dd666f

The *local* fixture **`ec2_provider_with_sts_creds`** is used in the following files:
- **cfme/tests/cloud/test_providers.py**
    - *test_add_ec2_provider_with_sts_assume_role*

Please, consider creating a PRT run to make sure your fixture changes do not break existing usage :smiley:

File: cfme/tests/cloud/test_providers.py [hunk 3]
```diff
@@ -186,6 +189,27 @@ def vm_ip(cfme_vhd, pwsh_ssh):
         pwsh_ssh.run_command("pwsh {}".format(path_script), timeout=180)
 
 
+@pytest.fixture
+def ec2_provider_with_sts_creds(appliance):
+    assume_role_creds = cfme_data['management_systems']['ec2west']['sts_assume_role']['credentials']
```
Comment (john-dupuy) [inline cfme/tests/cloud/test_providers.py]: If you first instantiate the provider object
```suggestion
    assume_role_creds = provider.data.sts_assume_role.credentials
```

File: cfme/tests/cloud/test_providers.py [hunk 4]
```diff
@@ -186,6 +189,27 @@ def vm_ip(cfme_vhd, pwsh_ssh):
...
+    creds = Credential(principal=credentials[assume_role_creds]['username'],
+                       secret=credentials[assume_role_creds]['password'])
+    endpoint = EC2Endpoint(
+        assume_role_arn=cfme_data['management_systems']['ec2west']['sts_assume_role']['role_arn'],
```
Comment (john-dupuy) [inline cfme/tests/cloud/test_providers.py]: Same here, should be accessed through `provider.data`

File: cfme/tests/cloud/test_providers.py [hunk 5]
```diff
@@ -186,6 +189,27 @@ def vm_ip(cfme_vhd, pwsh_ssh):
         pwsh_ssh.run_command("pwsh {}".format(path_script), timeout=180)
 
 
+@pytest.fixture
+def ec2_provider_with_sts_creds(appliance):
```
Comment (john-dupuy) [inline cfme/tests/cloud/test_providers.py]: I think this fixture can be simplified by first instantiating the ec2west provider and then just overriding it's endpoint attributes. 

```python,
provider = providers.get_crud("ec2west")
assume_role_creds = provider.data.sts_assume_role.credentials
creds = Credential(principal=credentials[assume_role_creds]['username'],
                       secret=credentials[assume_role_creds]['password'])
endpoint = EC2Endpoint(
        assume_role_arn=provider.data.sts_assume_role.role_arn,
        credentials=creds)
provider.endpoints = endpoint
```

What do you think?

Comment (mmojzis) [reply cfme/tests/cloud/test_providers.py]: Reworked

File: cfme/tests/cloud/test_providers.py (hunk 4 above)
Comment (mmojzis) [reply cfme/tests/cloud/test_providers.py]: Reworked

**Pull request merged**